### ISSUE TITLE
Cache coherency issues in related to the lack of multi-document atomicity

### ISSUE THREAD

**Issue by paulocsanz** [RocketChat/Rocket.Chat#22744]
### Description:

Since MongoDB is not very good at joins/lookups, projects have to bake caches of the referenced data and update it according to the real data chance. But this is impossible to ensure without multi-document atomicity, something mongo 4.x started supporting.

So with time, with downscales, upscales, deploys, crashes and any other downtime .Those caches may get invalidated and never fixed again. One example is `rocketchat_room.usersCount`. But it's widespread, virtually every cached structure is vulnerable to this problem. This causes usability issues in the long run.

We use a very old Rocket.Chat version, so I'm not asking for support for it, I'm asking that future versions start getting more careful about cache coherency issues. And I'm willing to work together on this groundwork to support mongo multi-document transactions inside Rocket.Chat's infrastructure. I just have no idea on how to start it right now.

### Steps to reproduce:

This function:  https://github.com/RocketChat/Rocket.Chat/blob/848cc13762431f0ce968f7d766bc599de0d5a065/app/models/server/models/Subscriptions.js#L1223-L1236

If between lines 1233 and 1236 the machine of the server shuts down, the cache will be invalid, forever. Because it only gets incremented/decremented, never actually checked against Subscriptions. The only way to ensure is using [mongo multi-document transactions](https://docs.mongodb.com/manual/core/transactions/). To make something like this:

```js
const client = new MongoClient(uri);
await client.connect();
const session = client.startSession();
await session.withTransaction(async () => {
  const Rooms = client.db('rocketchat').collection('rocketchat_rooms');
  const Subscriptions = client.db('rocketchat').collection('rocketchat_subscriptions');

  await Rooms.incUsersCountNotDMsByIds(roomIds, -1, { session });
  await Users.removeAllRoomsByUserId(userId, { session }); 
}, {
  readPreference: 'primary',
  readConcern: { level: 'local' },
  writeConcern: { w: 'majority' }
});
```

I have no idea how can this be feasible with Rocket.Chat Meteor models usage, maybe with Raw models, but it's not clear if the change to raw models can be done without impact. Neither how to extend the model API to get this session parameter.

Since it's a race-condition it's very hard to reproduce, we can only explain this with natural language with the distributed systems principles in mind. And maybe running a check for the entire database of other systems, check the subscriptions and the usersCount and how many don't match. We may have a script for that to provide if needed.

A quick win could also be a background task that fixes all the caches, and monitors how this cache breakage is affecting the user. It's not ideal, nor how I would like it fixed, but it can be a good product decision.

I also don't know how deep it affects the usability, if some features depends on this cached data to properly function and are getting broken. We saw that happen in our system, because we check those caches for decision-making, but I really don't have more documented/tangible information to give.

### Expected behavior:

Independently of when the shutdown happened, the caches must be coherent otherwise this lookup avoidance hurts usability.

### Actual behavior:

Some shutdowns cause cache incoherence through the code, I cited one small example, but there are others.

**Comment by johncrisp** [comment]
Hi Paulo and thanks for reporting as per our conversation.

Someone will take a look in due course.

**Comment by fdellwing** [comment]
I'm with you, that this is a problem and should be looked at. But

```
{
    readPreference: 'primary',
    readConcern: { level: 'local' },
    writeConcern: { w: 'majority' }
  });
```

this should not be present in the code? I want to have full control over this with my URI that I use to connect to the database. This might just be an artifact in your code snippet, but is important for us.

**Comment by paulocsanz** [comment]
Of course, this is not necessarily the best solution. I just copied it from the mongo multi-document transaction example.

We also control it with the URI, but I'm not sure if transactions have a different consistency requirement than the rest of the code. It seems something we need to understand better about mongo. And even if there is, maybe the user wants to ignore that consistency, so yeah it shouldn't be hardcoded here.
